pyro/ops/contract.py

```diff
@@ -495,7 +491,10 @@ def contract_to_tensor(tensor_tree, sum_dims, target_ordinal=None, target_dims=N
     contracted_terms = []
 
     # Split this tensor tree into connected components.
+    modulo_total = bool(target_dims)
```

Comment:
@martinjankowiak the following few lines are the only new logic required for optimization.